pkg/ccl/logictestccl/tests/fakedist-disk/fakedist-disk_test: TestCCLLogic_show_create failed

pkg/ccl/logictestccl/tests/fakedist-disk/fakedist-disk_test.TestCCLLogic_show_create [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7698310?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7698310?buildTab=artifacts#/) on master @ [1a6e9f885baa124d5ff2996adb966ea15a1a9b2b](https://github.com/cockroachdb/cockroach/commits/1a6e9f885baa124d5ff2996adb966ea15a1a9b2b):


```
=== RUN   TestCCLLogic_show_create
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/effab48683274210e068775452544fff/logTestCCLLogic_show_create2055064146
    test_log_scope.go:79: use -show-logs to present logs inline
*
* INFO: Running test with the default test tenant. If you are only seeing a test case failure when this message appears, there may be a problem with your test case running within tenants.
*
*
* INFO: Running test with the default test tenant. If you are only seeing a test case failure when this message appears, there may be a problem with your test case running within tenants.
*
*
* INFO: Running test with the default test tenant. If you are only seeing a test case failure when this message appears, there may be a problem with your test case running within tenants.
*
[06:30:34] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/2360/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/logictestccl/tests/fakedist-disk/fakedist-disk_test_/fakedist-disk_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/logictestccl/testdata/logic_test/show_create with config fakedist-disk: 8 tests, 0 failures
[06:30:35] --- total progress: 8 statements
--- total: 8 tests, 0 failures
    logic.go:4129: -- test log scope end --
    generated_test.go:108: Leaked goroutine: goroutine 44112 [select]:
        github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).startTickerLocked.func1()
        	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:302 +0xe5
        created by github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).startTickerLocked
        	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:297 +0x7a
--- FAIL: TestCCLLogic_show_create (8.16s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCCLLogic_show_create.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/settings/registry.go
+++ b/pkg/settings/registry.go
@@ -156,6 +156,7 @@ var retiredSettings = map[string]struct{}{
 	"sql.catalog.descs.validate_on_write.enabled": {},
 	"sql.distsql.max_running_flows":               {},
 	"sql.distsql.flow_scheduler_queueing.enabled": {},
+	"sql.distsql.drain.cancel_after_wait.enabled": {},
 }
 
 // sqlDefaultSettings is the list of "grandfathered" existing sql.defaults

--- a/pkg/sql/distsql/BUILD.bazel
+++ b/pkg/sql/distsql/BUILD.bazel
@@ -12,7 +12,6 @@ go_library(
         "//pkg/kv",
         "//pkg/roachpb",
         "//pkg/server/telemetry",
-        "//pkg/settings",
         "//pkg/sql/catalog/catsessiondata",
         "//pkg/sql/catalog/descs",
         "//pkg/sql/colflow",

--- a/pkg/sql/distsql/server.go
+++ b/pkg/sql/distsql/server.go
@@ -20,7 +20,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/server/telemetry"
-	"github.com/cockroachdb/cockroach/pkg/settings"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catsessiondata"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
 	"github.com/cockroachdb/cockroach/pkg/sql/colflow"
@@ -139,15 +138,6 @@ func (ds *ServerImpl) NumRemoteRunningFlows() int {
 	return ds.remoteFlowRunner.NumRunningFlows()
 }
 
-// TODO(yuzefovich): remove this setting in 23.1.
-var cancelRunningQueriesAfterFlowDrainWait = settings.RegisterBoolSetting(
-	settings.TenantWritable,
-	"sql.distsql.drain.cancel_after_wait.enabled",
-	"determines whether queries that are still running on a node being drained "+
-		"are forcefully canceled after waiting the 'server.shutdown.query_wait' period",
-	true,
-)
-
 // Drain changes the node's draining state through gossip and drains the
 // server's flowRegistry. See flowRegistry.Drain for more details.
 func (ds *ServerImpl) Drain(
@@ -167,8 +157,7 @@ func (ds *ServerImpl) Drain(
 		// wait a minimum time for the draining state to be gossiped.
 		minWait = 0
 	}
-	cancelStillRunning := cancelRunningQueriesAfterFlowDrainWait.Get(&ds.Settings.SV)
-	ds.flowRegistry.Drain(flowWait, minWait, reporter, cancelStillRunning)
+	ds.flowRegistry.Drain(flowWait, minWait, reporter)
 }
 
 // setDraining changes the node's draining state through gossip to the provided

--- a/pkg/sql/distsql/setup_flow_after_drain_test.go
+++ b/pkg/sql/distsql/setup_flow_after_drain_test.go
@@ -47,8 +47,7 @@ func TestSetupFlowAfterDrain(t *testing.T) {
 		remoteFlowRunner,
 	)
 	distSQLSrv.flowRegistry.Drain(
-		time.Duration(0) /* flowDrainWait */, time.Duration(0), /* minFlowDrainWait */
-		nil /* reporter */, false, /* cancelStillRunning */
+		time.Duration(0) /* flowDrainWait */, time.Duration(0) /* minFlowDrainWait */, nil, /* reporter */
 	)
 
 	// We create some flow; it doesn't matter what.

--- a/pkg/sql/flowinfra/flow_registry.go
+++ b/pkg/sql/flowinfra/flow_registry.go
@@ -456,7 +456,6 @@ func (fr *FlowRegistry) Drain(
 	flowDrainWait time.Duration,
 	minFlowDrainWait time.Duration,
 	reporter func(int, redact.SafeString),
-	cancelStillRunning bool,
 ) {
 	allFlowsDone := make(chan struct{}, 1)
 	start := timeutil.Now()
@@ -483,16 +482,14 @@ func (fr *FlowRegistry) Drain(
 			time.Sleep(expectedConnectionTime)
 			fr.Lock()
 		}
-		if cancelStillRunning {
-			// Now cancel all still running flows.
-			for _, f := range fr.flows {
-				if f.flow != nil && f.flow.ctxCancel != nil {
-					// f.flow might be nil when ConnectInboundStream() was
-					// called, but the consumer of that inbound stream hasn't
-					// been scheduled yet.
-					// f.flow.ctxCancel might be nil in tests.
-					f.flow.ctxCancel()
-				}
+		// Now cancel all still running flows.
+		for _, f := range fr.flows {
+			if f.flow != nil && f.flow.ctxCancel != nil {
+				// f.flow might be nil when ConnectInboundStream() was
+				// called, but the consumer of that inbound stream hasn't
+				// been scheduled yet.
+				// f.flow.ctxCancel might be nil in tests.
+				f.flow.ctxCancel()
 			}
 		}
 		fr.Unlock()

--- a/pkg/sql/flowinfra/flow_registry_test.go
+++ b/pkg/sql/flowinfra/flow_registry_test.go
@@ -393,7 +393,7 @@ func TestFlowRegistryDrain(t *testing.T) {
 		registerFlow(t, id)
 		drainDone := make(chan struct{})
 		go func() {
-			reg.Drain(math.MaxInt64 /* flowDrainWait */, 0 /* minFlowDrainWait */, nil /* reporter */, false /* cancelStillRunning */)
+			reg.Drain(math.MaxInt64 /* flowDrainWait */, 0 /* minFlowDrainWait */, nil /* reporter */)
 			drainDone <- struct{}{}
 		}()
 		// Be relatively sure that the FlowRegistry is draining.
@@ -406,7 +406,7 @@ func TestFlowRegistryDrain(t *testing.T) {
 	// DrainTimeout verifies that Drain returns once the timeout expires.
 	t.Run("DrainTimeout", func(t *testing.T) {
 		registerFlow(t, id)
-		reg.Drain(0 /* flowDrainWait */, 0 /* minFlowDrainWait */, nil /* reporter */, false /* cancelStillRunning */)
+		reg.Drain(0 /* flowDrainWait */, 0 /* minFlowDrainWait */, nil /* reporter */)
 		reg.UnregisterFlow(id)
 		reg.Undrain()
 	})
@@ -417,7 +417,7 @@ func TestFlowRegistryDrain(t *testing.T) {
 		registerFlow(t, id)
 		drainDone := make(chan struct{})
 		go func() {
-			reg.Drain(math.MaxInt64 /* flowDrainWait */, 0 /* minFlowDrainWait */, nil /* reporter */, false /* cancelStillRunning */)
+			reg.Drain(math.MaxInt64 /* flowDrainWait */, 0 /* minFlowDrainWait */, nil /* reporter */)
 			drainDone <- struct{}{}
 		}()
 		// Be relatively sure that the FlowRegistry is draining.
@@ -460,7 +460,7 @@ func TestFlowRegistryDrain(t *testing.T) {
 		}
 		defer func() { reg.testingRunBeforeDrainSleep = nil }()
 		go func() {
-			reg.Drain(math.MaxInt64 /* flowDrainWait */, 0 /* minFlowDrainWait */, nil /* reporter */, false /* cancelStillRunning */)
+			reg.Drain(math.MaxInt64 /* flowDrainWait */, 0 /* minFlowDrainWait */, nil /* reporter */)
 			drainDone <- struct{}{}
 		}()
 		if err := <-errChan; err != nil {
@@ -488,7 +488,7 @@ func TestFlowRegistryDrain(t *testing.T) {
 		minFlowDrainWait := 10 * time.Millisecond
 		start := timeutil.Now()
 		go func() {
-			reg.Drain(math.MaxInt64 /* flowDrainWait */, minFlowDrainWait, nil /* reporter */, false /* cancelStillRunning */)
+			reg.Drain(math.MaxInt64 /* flowDrainWait */, minFlowDrainWait, nil /* reporter */)
 			drainDone <- struct{}{}
 		}()
 		// Be relatively sure that the FlowRegistry is draining.

--- a/pkg/sql/flowinfra/remote_flow_runner.go
+++ b/pkg/sql/flowinfra/remote_flow_runner.go
@@ -70,6 +70,8 @@ func (r *RemoteFlowRunner) Init(metrics *execinfra.DistSQLMetrics) {
 
 // RunFlow starts the given flow; does not wait for the flow to complete.
 func (r *RemoteFlowRunner) RunFlow(ctx context.Context, f Flow) error {
+	// cleanedUp is only accessed from the current goroutine.
+	cleanedUp := false
 	err := r.stopper.RunTaskWithErr(
 		ctx, "flowinfra.RemoteFlowRunner: running flow", func(ctx context.Context) error {
 			log.VEventf(ctx, 1, "flow runner running flow %s", f.GetID())
@@ -93,7 +95,6 @@ func (r *RemoteFlowRunner) RunFlow(ctx context.Context, f Flow) error {
 				f.Cleanup(ctx)
 				return err
 			}
-			// The flow can be started.
 			r.metrics.FlowStart()
 			cleanup := func() {
 				func() {
@@ -105,18 +106,40 @@ func (r *RemoteFlowRunner) RunFlow(ctx context.Context, f Flow) error {
 				r.metrics.FlowStop()
 				f.Cleanup(ctx)
 			}
-			if err := f.Start(ctx); err != nil {
-				cleanup()
-				return err
-			}
-			go func() {
+			// First, make sure that we can spin up a new async task whose job
+			// is to wait for the flow to finish and perform the cleanup.
+			//
+			// However, we need to make sure that this task blocks until the
+			// flow is started. True value will be sent on waiterShouldExit if
+			// we couldn't start the flow and the async task must exit right
+			// away, without waiting; when the channel is closed, the flow has
+			// been started successfully, and the async task proceeds to waiting
+			// for its completion.
+			waiterShouldExit := make(chan bool)
+			if err := r.stopper.RunAsyncTask(ctx, "flowinfra.RemoteFlowRunner: waiting for flow to finish", func(ctx context.Context) {
+				if shouldExit := <-waiterShouldExit; shouldExit {
+					return
+				}
 				f.Wait()
 				cleanup()
-			}()
+			}); err != nil {
+				cleanup()
+				cleanedUp = true
+				return err
+			}
+			// Now, start the flow to run concurrently.
+			if err := f.Start(ctx); err != nil {
+				cleanup()
+				cleanedUp = true
+				waiterShouldExit <- true
+				return err
+			}
+			close(waiterShouldExit)
 			return nil
 		})
-	if err != nil && errors.Is(err, stop.ErrUnavailable) {
-		// If the server is quiescing, we have to explicitly clean up the flow.
+	if err != nil && errors.Is(err, stop.ErrUnavailable) && !cleanedUp {
+		// If the server is quiescing, we have to explicitly clean up the flow
+		// if it hasn't been cleaned up yet.
 		f.Cleanup(ctx)
 	}
 	return err

